cluster_management/src/main/java/com/pinterest/rocksplicator/eventstore/LeaderEventHistoryStore.java

```diff
@@ -0,0 +1,498 @@
+    try {
+      this.zkClient.blockUntilConnected(60, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      LOGGER.error(String.format("Can't connect to zookeeper: %s", zkConnectString), e);
+    }
+
+    this.zkStoreCache = CacheBuilder.newBuilder()
+        .expireAfterAccess(10, TimeUnit.MINUTES)
+        .removalListener(
+            new RemovalListener<String, LoadingCache<String,
+                MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent>>>() {
+              @Override
+              public void onRemoval(
+                  RemovalNotification<String, LoadingCache<String,
+                      MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent>>> notification) {
+                notification.getValue().invalidateAll();
+              }
+            })
+        .build(new CacheLoader<String,
+            LoadingCache<String, MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent>>>() {
+          @Override
+          public LoadingCache<String, MergeableReadWriteStore<LeaderEventsHistory,
+              LeaderEvent>> load(
+              String resourceName) throws Exception {
+            return CacheBuilder.newBuilder()
+                .expireAfterAccess(10, TimeUnit.MINUTES)
+                .removalListener(
+                    new RemovalListener<String, MergeableReadWriteStore<LeaderEventsHistory,
+                        LeaderEvent>>() {
+                      @Override
+                      public void onRemoval(
+                          RemovalNotification<String, MergeableReadWriteStore<LeaderEventsHistory,
+                              LeaderEvent>> notification) {
+                        try {
+                          notification.getValue().close();
+                        } catch (IOException e) {
+                          LOGGER.warn("Error while closing the zkStore", e);
+                        }
+                      }
+                    })
+                .build(new CacheLoader<String,
+                    MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent>>() {
+                  @Override
+                  public MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent> load(
+                      String partitionName) throws Exception {
+                    return new ZkMergeableEventStore<>(zkClient, clusterName,
+                        resourceName, partitionName, leaderEventsHistoryCodec,
+                        LeaderEventsHistory::new,
+                        MergeOperators
+                            .createBatchMergeOperator(resourceName, partitionName,
+                                maxEventsToKeep));
+                  }
+                });
+          }
+        });
+
+    this.executorServices = new ArrayList<>(NUM_PARALLEL_THREADS);
+
+    for (int i = 0; i < NUM_PARALLEL_THREADS; ++i) {
+      this.executorServices.add(Executors.newSingleThreadExecutor());
+    }
+    this.clusterResourcePartitionLock = new ArrayList<>(128);
```

Comment:
Why 128?

Reply:
This should have been same as number of parallel threads used for updating zk. We use SingleThreadExecutor... hence we select which executor should run a given task of updating the zk leader events history for given resource.